oktaawscli/aws_auth.py

```diff
@@ -52,19 +55,23 @@ def choose_aws_role(self, assertion):
 of roles assigned to you.""" % self.role)
                 self.logger.info("Please choose a role.")
 
+        if len(roles) == 1:
+            sys.stderr.write("\nOne role found, using role: " + roles[0][1] + "\n")
+            return role_info[0]
+
         role_options = self.__create_options_from(role_info)
         role_choice = None
         while role_choice is None:
             try:
                 for option in role_options:
-                    print(option)
-
-                role_choice = int(input('Please select the AWS role: ')) - 1
+                    sys.stderr.write(option + "\n")
+                sys.stderr.write('Please select the AWS role: ')
+                role_choice = int(input()) - 1
                 if role_choice >= 0 and role_choice < len(role_info):
                     return role_info[role_choice]
                 raise IndexError('Bad selection')
-            except (SyntaxError, NameError, ValueError, IndexError) as ex:
-                print("\nYou have selected an invalid role index, please try again.\n")
+            except (SyntaxError, NameError, ValueError, IndexError):
+                sys.stderr.write("\nYou have selected an invalid role index, please try again.\n")
```

Comment:
Two things:
Why do we want an interactive script to work entirely over stderr? Stderr is typically reserved for error messages or other exceptional cases and we shouldn't use it primarily.
Rather than deal with the differences between `print` and `sys.stderr.write` and newlines and formatting and such, we should import `from __future__ import print_function` and use the `file` argument to redirect print to stderr.
Example: `print("some error message", file=sys.stderr)`

Reply:
Similar to the original repo, stderr was used as a workaround to still permit user input and feedback when using backquotes for command substitution since stdout is what's used in the substitution which we intended to use to have the export option with the "-e" export flag export it to our parent shell in one shot (our first tool "awssudo" used to be used this way). I'm thinking an alternative solution not involving stderr would be to actually have the keys/secrets exported into the env of the script's runtime and the parent shell to call it using "source"